### PR TITLE
Provide the option to always regenerate the jsps

### PR DESCRIPTION
**Pull request by jadelkhoury**
solves #128

Introducing a doNotRegenerateJspFiles attribute per adaptor.
If enabled, jsp files are not generated if they exist. Missing files are however still generated.

Removing all protected user code blocks from jsp-files. For existing models, this will cause existing content of jsp files to be placed in *.lost files.

### CODE REVIEW

File: org.eclipse.lyo.oslc4j.adaptormodel/model/vocabulary.ecore [hunk 1]
```diff
@@ -8,12 +8,7 @@
   </eClassifiers>
   <eClassifiers xsi:type="ecore:EClass" name="Vocabulary">
     <eStructuralFeatures xsi:type="ecore:EAttribute" name="namespaceURI" lowerBound="1"
-        eType="ecore:EDataType http://www.eclipse.org/emf/2002/Ecore#//EString">
```
Comment (berezovskyi) [inline org.eclipse.lyo.oslc4j.adaptormodel/model/vocabulary.ecore]: what's this change about?

File: org.eclipse.lyo.oslc4j.codegenerator/src/org/eclipse/lyo/oslc4j/codegenerator/services/fileServices.mtl [hunk 2]
```diff
@@ -0,0 +1,10 @@
+[comment encoding = UTF-8 /]
+[module fileServices('http://org.eclipse.lyo/oslc4j/adaptorInterface')/]
+
+[query public setTargetFolder(path : String): Boolean
```
Comment (berezovskyi) [inline org.eclipse.lyo.oslc4j.codegenerator/src/org/eclipse/lyo/oslc4j/codegenerator/services/fileServices.mtl]: that's not really a query but a side-effect

**Comment by jadelkhoury** [comment]
@jamsden I would like your input on this request, since it affects an old change you requested to allow full flexibility to change the content of the jsp-pages.
The approach proposed in this pull request is much neater than the current approach. A flag can now be set/unset to decide if one should regenerate the jsp pages. There is no longer a need for the protected user blocks.
The problem? For existing generations, all content in jsp-files between the protected user blocks is moved to *.lost files. So, it is an annoying transient change for current models.

File: org.eclipse.lyo.oslc4j.adaptormodel/model/vocabulary.ecore (hunk 1 above)
Comment (jadelkhoury) [reply org.eclipse.lyo.oslc4j.adaptormodel/model/vocabulary.ecore]: Just give me 2 minutes to finish writing :-)

**Comment by jadelkhoury** [comment]
Most files in the pull request are generated. So, 
1. filter out the java files (https://github.com/eclipse/lyo.designer/pull/132/files?file-filters%5B%5D=.MF&file-filters%5B%5D=.ecore&file-filters%5B%5D=.genmodel&file-filters%5B%5D=.mtl&file-filters%5B%5D=.properties)

2. Review these 2 java files
* org.eclipse.lyo.oslc4j.codegenerator/src/org/eclipse/lyo/oslc4j/codegenerator/main/Generate.java
* org.eclipse.lyo.oslc4j.codegenerator/src/org/eclipse/lyo/oslc4j/codegenerator/services/FileServices.java

File: org.eclipse.lyo.oslc4j.adaptormodel/model/vocabulary.ecore (hunk 1 above)
Comment (jadelkhoury) [reply org.eclipse.lyo.oslc4j.adaptormodel/model/vocabulary.ecore]: I had problems compiling/generating java from the ecore files due to this.
I guess new version of EMF became more picky, causing errors that we did not have before.

File: org.eclipse.lyo.oslc4j.codegenerator/src/org/eclipse/lyo/oslc4j/codegenerator/services/fileServices.mtl (hunk 2 above)
Comment (jadelkhoury) [reply org.eclipse.lyo.oslc4j.codegenerator/src/org/eclipse/lyo/oslc4j/codegenerator/services/fileServices.mtl]: yes. I needed to save the targetFolder, so I don't have to pass it around all over the acceleo code.

File: org.eclipse.lyo.tools.adaptormodel.editor/META-INF/MANIFEST.MF [hunk 3]
```diff
@@ -2,6 +2,7 @@ Manifest-Version: 1.0
 Bundle-ManifestVersion: 2
 Bundle-Name: %pluginName
 Bundle-SymbolicName: org.eclipse.lyo.tools.adaptormodel.editor;singleton:=true
+Automatic-Module-Name: org.eclipse.lyo.tools.adaptormodel.editor
```
Comment (berezovskyi) [inline org.eclipse.lyo.tools.adaptormodel.editor/META-INF/MANIFEST.MF]: these a Java 9 things, just FYI

File: org.eclipse.lyo.oslc4j.codegenerator/src/org/eclipse/lyo/oslc4j/codegenerator/services/jspServices.mtl [hunk 4]
```diff
@@ -24,6 +24,11 @@
 [import org::eclipse::lyo::oslc4j::codegenerator::services::adaptorInterfaceServices/]
 [import org::eclipse::lyo::oslc4j::codegenerator::services::resourceServices/]
 [import org::eclipse::lyo::oslc4j::codegenerator::services::serviceProviderServices/]
+[import org::eclipse::lyo::oslc4j::codegenerator::services::fileServices/]
+
+[query public generateJspFile(anAdaptorInterface : AdaptorInterface, filePath : String) : Boolean = 
+    not (anAdaptorInterface.doNotRegenerateJspFiles()._and(fileExists(filePath)))
```
Comment (berezovskyi) [inline org.eclipse.lyo.oslc4j.codegenerator/src/org/eclipse/lyo/oslc4j/codegenerator/services/jspServices.mtl]: **NB!** I think the logic proposed by @jposterloh was to refrain from generating code in `webapp-gen/a/b/c` if the developer already created `webapp/a/b/c`. So this requires a different target folder.

**Comment by jadelkhoury** [comment]
I assume @berezovskyi  that you still have an issue with the setter method in the java class? 

@jposterloh! The proposal here does not match your suggestion. I only generate to a single folder, but you can control whether files are generated or not. I think this is a bit simpler.

**Comment by berezovskyi** [comment]
@jadelkhoury, yes, I still do.

From http://docs.obeonetwork.com/obeodesigner/6.1/viewpoint/specifier/general/Writing_Queries.html#service_methods

> Warning: Java service methods should be stateless. There is no guarantee that two successive invocations of the same service method on two model elements (or even on the same one) will use the same instance of the service class.

**Comment by jadelkhoury** [comment]
@berezovskyi How about the suggested use of using a static variable (see 2nd commit).

Note that http://docs.obeonetwork.com/obeodesigner/6.1/viewpoint/specifier/general/Writing_Queries.html#service_methods refers to the use of Acceleo within a Sirius modelling project. This is not our scenario here, since we are using an Acceleo code generation. The more relevant rules are from https://wiki.eclipse.org/Acceleo/Getting_Started#Java_services_wrappers

Note for example that Sirius requires the first argument to be an EObject. Acceleo code generator does not.

But, I agree that generally speaking Acceleo queries are meant to be stateless. If you call an Acceleo query twice, Acceleo does not even it run it a second time, since it is meant to return the same result.
This we are guaranteeing here, since I only set the targetfolder once at the start.

**Pull request merged**